### PR TITLE
Fix: Create a container with Duplicate Name

### PR DESCRIPTION
**Pull request by yzxiu**
fix https://github.com/containerd/nerdctl/issues/1787

Signed-off-by: yaozhenxiu <946666800@qq.com>

### CODE REVIEW

**Comment by Zheaoli** [comment]
We may need extra e2e test for this case

File: pkg/namestore/namestore.go [hunk 1]
```diff
@@ -115,3 +132,21 @@ func (x *nameStore) Rename(oldName, id, newName string) error {
 	}
 	return lockutil.WithDirLock(x.dir, fn)
 }
+
+func GetPrintableContainerName(containerLabels map[string]string) string {
```
Comment (AkihiroSuda) [inline pkg/namestore/namestore.go:136]:
```suggestion
func PrintableContainerName(containerLabels map[string]string) string {
```

https://go.dev/doc/effective_go#Getters

Comment (yzxiu) [reply pkg/namestore/namestore.go]: thanks

File: pkg/namestore/namestore.go [hunk 2]
```diff
@@ -59,6 +63,19 @@ func (x *nameStore) Acquire(name, id string) error {
 		if b, err := os.ReadFile(fileName); err == nil {
 			return fmt.Errorf("name %q is already used by ID %q", name, string(b))
 		}
+		containers, err := client.Containers(ctx)
+		if err != nil {
+			return err
+		}
+		for _, c := range containers {
```
Comment (djdongjin) [inline pkg/namestore/namestore.go:70]: nit: for loop can be parallelized

```golang
        var eg errgroup.Group
		for _, c := range containers {
            c := c
             eg.Go(func() error {
                  // your for loop
             })
		}
	    if err := eg.Wait(); err != nil {
		  return err
	    }
```

Comment (yzxiu) [reply pkg/namestore/namestore.go:72]: thanks

File: cmd/nerdctl/create_linux_test.go [hunk 3]
```diff
@@ -109,6 +111,32 @@ func TestCreateWithMACAddress(t *testing.T) {
...
+	if len(store) != 2 {
+		t.Fatalf("parse store path fail")
+	}
+	// store[0]
+	// root: /var/lib/nerdctl/1935db59
+	// rootless: ~/.local/share/nerdctl/1935db59
+	fileName := fmt.Sprintf("%s/names/%s/%s", store[0], testutil.Namespace, containerName)
+	if _, err := os.ReadFile(fileName); err != nil {
+		t.Fatalf("container %s file name not exist", containerName)
+	}
```
Comment (AkihiroSuda) [inline cmd/nerdctl/create_linux_test.go:133]: assert.NilError

File: cmd/nerdctl/create_linux_test.go [hunk 4]
```diff
@@ -109,6 +111,32 @@ func TestCreateWithMACAddress(t *testing.T) {
...
+	imageName := testutil.CommonImage
+	containerName := testutil.Identifier(t)
+
+	base.Cmd("create", "--name", containerName, imageName).AssertOK()
+	defer base.Cmd("container", "rm", "-f", containerName).AssertOK()
+
+	store := strings.Split(base.InspectContainer(containerName).ResolvConfPath, "/containers/")
+	if len(store) != 2 {
+		t.Fatalf("parse store path fail")
+	}
```
Comment (AkihiroSuda) [inline cmd/nerdctl/create_linux_test.go:126]: assert.Assert

File: cmd/nerdctl/create_linux_test.go [hunk 5]
```diff
@@ -109,6 +111,32 @@ func TestCreateWithMACAddress(t *testing.T) {
 	}
 }
 
+func TestCreateDuplicateName(t *testing.T) {
```
Comment (AkihiroSuda) [inline cmd/nerdctl/create_linux_test.go:114]: Probably this test is not specific to Linux, and can be moved to `create_test.go`

Comment (AkihiroSuda) [reply cmd/nerdctl/create_linux_test.go:114]: Also could you add comment lines to explain this test scenario?

File: cmd/nerdctl/create_linux_test.go [hunk 6]
```diff
@@ -109,6 +111,32 @@ func TestCreateWithMACAddress(t *testing.T) {
...
+
+	store := strings.Split(base.InspectContainer(containerName).ResolvConfPath, "/containers/")
+	if len(store) != 2 {
+		t.Fatalf("parse store path fail")
+	}
+	// store[0]
+	// root: /var/lib/nerdctl/1935db59
+	// rootless: ~/.local/share/nerdctl/1935db59
+	fileName := fmt.Sprintf("%s/names/%s/%s", store[0], testutil.Namespace, containerName)
+	if _, err := os.ReadFile(fileName); err != nil {
```
Comment (AkihiroSuda) [inline cmd/nerdctl/create_linux_test.go:131]: os.Stat might be enough

File: pkg/namestore/namestore.go [hunk 7]
```diff
@@ -59,6 +64,27 @@ func (x *nameStore) Acquire(name, id string) error {
...
+		}
+		var eg errgroup.Group
+		for _, c := range containers {
+			c := c
+			eg.Go(func() error {
+				info, err := c.Info(ctx, containerd.WithoutRefreshedMetadata)
+				if err != nil {
+					return err
+				}
+				if PrintableContainerName(info.Labels) == name {
```
Comment (AkihiroSuda) [inline pkg/namestore/namestore.go:79]: Just checking `info.Labels[labels.Name]` should be enough

Comment (AkihiroSuda) [reply pkg/namestore/namestore.go:79]: (So, `getPrintableContainerName()` does not need to be moved in this PR; sorry for going back and forth)

File: cmd/nerdctl/create_test.go [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+func TestCreateDuplicateName(t *testing.T) {
+	testutil.DockerIncompatible(t)
+	base := testutil.NewBase(t)
+	imageName := testutil.CommonImage
+	containerName := testutil.Identifier(t)
+
+	base.Cmd("create", "--name", containerName, imageName).AssertOK()
+	defer base.Cmd("container", "rm", "-f", containerName).AssertOK()
+
+	logrus.Printf("inspect:%v", base.InspectContainer(containerName))
```
Comment (manugupt1) [inline cmd/nerdctl/create_test.go:41]: t.Log

File: cmd/nerdctl/create_test.go [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+	store := strings.Split(base.InspectContainer(containerName).ResolvConfPath, "/containers/")
+	logrus.Printf("store: %v", store)
+	if len(store) != 2 {
```
Comment (manugupt1) [inline cmd/nerdctl/create_test.go:44]: assert.Equal(2, len(store))

File: pkg/namestore/namestore.go [hunk 10]
```diff
@@ -59,6 +63,27 @@ func (x *nameStore) Acquire(name, id string) error {
...
+		var eg errgroup.Group
+		for _, c := range containers {
+			c := c
+			eg.Go(func() error {
+				info, err := c.Info(ctx, containerd.WithoutRefreshedMetadata)
+				if err != nil {
+					return err
+				}
+				if getContainerName(info.Labels) == name {
+					return fmt.Errorf("name %q is already used by ID %q", name, info.ID)
```
Comment (manugupt1) [inline pkg/namestore/namestore.go:79]: Can you please update the error message to ask user to remove the container using `rm -f`

File: cmd/nerdctl/create_test.go [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+	base := testutil.NewBase(t)
+	imageName := testutil.CommonImage
+	containerName := testutil.Identifier(t)
+
+	base.Cmd("create", "--name", containerName, imageName).AssertOK()
+	defer base.Cmd("container", "rm", "-f", containerName).AssertOK()
+
+	logrus.Printf("inspect:%v", base.InspectContainer(containerName))
+	store := strings.Split(base.InspectContainer(containerName).ResolvConfPath, "/containers/")
+	logrus.Printf("store: %v", store)
```
Comment (manugupt1) [inline cmd/nerdctl/create_test.go:43]: t.Log

**Comment by yzxiu** [comment]
@Zheaoli @manugupt1 @AkihiroSuda @djdongjin    Thanks for review.

This problem seems to be caused by my mistaken deletion of files.
I think these changes are not necessary, so I close it...

**Pull request closed without merging**